### PR TITLE
HW#4_Trupkina

### PR DESCRIPTION
**Pull request by SvetlanaTr**

### CODE REVIEW

File: HW4/HW4_SvTr.js [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+// and fix them with: 
+// - call, apply, bind
+// - arrow functions
+const firstObj = {
+    objHeight: 175,
+    sayHeigth: function () {
+        return this.objHeight + ' good!';
+    }
+};
+//var getH = firstObj.sayHeight;  //undefined (value 'this' have a meaning windows)
```
Comment (DmitryKorolDevPro) [inline HW4/HW4_SvTr.js:11]: pls, do double check

File: HW4/HW4_SvTr.js [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+//getH(); //TypeError
+firstObj.sayHeigth.call(firstObj);//175 good! ('call' set value 'this' = firstObj)
+firstObj.sayHeigth.apply(firstObj);//175 good! (also)
+var getH = firstObj.sayHeigth.bind(firstObj); //(to bind context 'this' with func firstObj)
```
Comment (DmitryKorolDevPro) [inline HW4/HW4_SvTr.js:15]: firstObj - in this case it is not indicative. without it will work the same way

File: HW4/HW4_SvTr.js [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+getH();// 175 good!
+const sayHh = () => firstObj.sayHeigth();//directly set value 'yhis' context 
+sayHh(); //175 good!
+
+//1.2 
+Math.max(3, 6, 1, 8, 2);//8
+var arrFind = [3, 6, 1, 8, 2];
+Math.max(arrFind);// lose this because Max leads to number array
+Math.max.apply(Math, arrFind);
```
Comment (DmitryKorolDevPro) [inline HW4/HW4_SvTr.js]: what exactly are you trying to show here?

File: HW4/HW4_SvTr.js [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    ],
+    entity_id: "someId1",
+}
+
+function copy(mainObj) {
+    let objCopy = {}; // objCopy will store mainObj
+    let key;
+
+    for (key in mainObj) {
+        objCopy[key] = mainObj[key]; // copy all properties into objCopy
```
Comment (DmitryKorolDevPro) [inline HW4/HW4_SvTr.js:94]: what if there are nested objects?
it will copy by reference, not by value

File: HW4/HW4_SvTr.js [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+console.log(copy(menu));
+/*
+{0: {…}}
+0: {product_name: "burger menu", tax: 0.5, subtotal: 4.5, total: 5, items: Array(1), …}
+__proto__: Object
+ */
+
+
+// 2.2 Spread {... obj}
+var cloneMenu1 = { ...menu };
```
Comment (DmitryKorolDevPro) [inline HW4/HW4_SvTr.js:107]: the same question about nested objects

File: HW4/HW4_SvTr.js [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+console.log(cloneMenu1);
+/**
+ {0: {…}}
+0: {product_name: "burger menu", tax: 0.5, subtotal: 4.5, total: 5, items: Array(1), …}
+__proto__: Object
+ */
+
+
+//2.3 Object.assign({}, obj)
+const cloneMenu2 = Object.assign({}, menu);
```
Comment (DmitryKorolDevPro) [inline HW4/HW4_SvTr.js:117]: Object.assign() performs a shallow copy of an object, not a deep clone.

File: HW4/HW4_SvTr.js [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+}];
+
+var menuItems = new Map();
+
+getMapNoDublicate(menu3);
+
+function getMapNoDublicate(data) {
+    for (let i = 0; i < data.length; i++) {
+        menuItems.set(data[i].id, data[i]);//method map.set() - add object if it does not exist (OR)
+        debugger;                          //                 - if it exist then updates element
```
Comment (DmitryKorolDevPro) [inline HW4/HW4_SvTr.js]: why did you leave it here?)

File: HW4/HW4_SvTr.js [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+    }]
+}];
+
+var menuItems = new Map();
+
+getMapNoDublicate(menu3);
+
+function getMapNoDublicate(data) {
+    for (let i = 0; i < data.length; i++) {
+        menuItems.set(data[i].id, data[i]);//method map.set() - add object if it does not exist (OR)
```
Comment (DmitryKorolDevPro) [inline HW4/HW4_SvTr.js:171]: what happens if we call this function several times in a row?

File: HW4/HW4_SvTr.js [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+        debugger;                          //                 - if it exist then updates element
+        //i can use iy like a checking
```
Comment (DmitryKorolDevPro) [inline HW4/HW4_SvTr.js]: please double-check the text you write

File: HW4/HW4_SvTr.js [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+        if (data[i].items && data[i].items.length) {//i check if element 'items' exist(true) and
```
Comment (DmitryKorolDevPro) [inline HW4/HW4_SvTr.js]: in modern js you can use data[i].items?.length

File: HW4/HW4_SvTr.js [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+    for (let i = 0; i < data.length; i++) {
+        menuItems.set(data[i].id, data[i]);//method map.set() - add object if it does not exist (OR)
+        debugger;                          //                 - if it exist then updates element
+        //i can use iy like a checking
+        if (data[i].items && data[i].items.length) {//i check if element 'items' exist(true) and 
+            getMapNoDublicate(data[i].items);//recursively calls for nested 'items'
+        }
+    }
+
+    return false;
```
Comment (DmitryKorolDevPro) [inline HW4/HW4_SvTr.js]: a function called "getMapNoDublicate" should not return false

File: HW4/HW4_SvTr.js (hunk 10 above)
Comment (SvetlanaTr) [reply HW4/HW4_SvTr.js]:
```
 if (data[i].items) {
            getMapNoDublicate(data[i].items);
        }
```
if i change like that? What do you think?

**Pull request closed without merging**